component/component.go

```diff
@@ -40,4 +56,17 @@ type Host interface {
 	// Context returns a context provided by the host to be used on the component
 	// operations.
 	Context() context.Context
+
+	// GetFactories of the specified kind. Returns a map of component types to factories.
+	// This allows components to create other components. For example:
+	//   func (r MyReceiver) Start(host component.Host) error {
+	//     factories := host.GetFactories(KindReceiver)
+	//     apacheFactory := factories["apache"].(receiver.Factory)
+	//     receiver, err := apacheFactory.CreateMetricsReceiver(...)
+	//     ...
```

Comment:
It seems a bit odd to ask user to cast to the factory kind when I passed the kind to the call. Can we have maybe separate calls `func GetReceiverFactories() map[string]receiver.Factory`?

Reply:
The problem is that `receiver.Factory` is in a package that imports this package so it will create a circular dependency.
If we want to have strongly-type return here we will need to move all Factory type to either this package or a separate package that can be imported from here.
The approach in this PR is the minimally invasive way to introduce this capability but if you feel that it is not nice enough we will have to do deeper refactoring and move more stuff around.

Reply:
@bogdandrutu any thoughts on this?

Reply:
Can you evaluate how hard is to remove that circular dependency?